Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,812 @@
+    isImplicit   = GHC.ideclImplicit idecl'
+    qualifier    = unpackFS <$> ghc_sl_fs <$> GHC.ideclPkgQual idecl'
+    hiding       = (catMaybes . parseHiding . ghcIdeclHiding) idecl'
+    importedAs   = (showSDoc dflags . ppr) <$> ideclAs idecl'
+    specifically = (parseSpecifically . ghcIdeclHiding) idecl'
+
+    grabNames :: GHC.Located [GHC.LIE GHC.RdrName] -> [String]
+    grabNames loc = map (showSDoc dflags . ppr) names
+      where names :: [RdrName]
+            names = map (ieName . SrcLoc.unLoc) $ SrcLoc.unLoc loc
+            -- FIXME We are throwing away location info by using unLoc each time?
+            -- Trace these things to see what we are losing.
+
+    parseHiding :: Maybe (Bool, Located [LIE RdrName]) -> [Maybe String]
+    parseHiding Nothing = [Nothing]
+
+    -- If we do
+    --
+    --     import System.Environment ( getArgs )
+    --
+    -- then we get ["getArgs"] here, but we don't really need it...
+    parseHiding (Just (False, _)) = []
+
+    -- Actually hid names, e.g.
+    --
+    --     import Data.List hiding (map)
+    parseHiding (Just (True, h))  = map Just $ grabNames h
+
+    parseSpecifically :: Maybe (Bool, Located [LIE RdrName]) -> [String]
+    parseSpecifically (Just (False, h)) = grabNames h
+    parseSpecifically _                 = []
+
+-- | Returns True if the 'Symbol' matches the end of the 'QualifiedName'.
+--
+-- Example:
+--
+-- >>> postfixMatch "bar" "Foo.bar"
+-- True
+-- >>> postfixMatch "bar" "Foo.baz"
+-- False
+-- >>> postfixMatch "bar" "bar"
+-- True
+postfixMatch :: String -> QualifiedName -> Bool
+postfixMatch originalSymbol qName = endTerm `isSuffixOf` qName
+  where endTerm = Safe.lastNote
+                    ("postfixMatch: got: " ++ originalSymbol)
+                    (splitOn "." originalSymbol)
+
+-- | Get the module part of a qualified name.
+--
+-- Example:
+--
+-- >>> moduleOfQualifiedName "Foo.bar"
+-- Just "Foo"
+-- >>> moduleOfQualifiedName "Foo"
+-- Nothing
+moduleOfQualifiedName :: QualifiedName -> Maybe String
+moduleOfQualifiedName qn = if null bits
+                                then Nothing
+                                else Just $ intercalate "." bits
+  where bits = reverse $ drop 1 $ reverse $ splitOn "." qn
+
+-- | Find the possible qualified names for the symbol at line/col in the given Haskell file and module.
+-- Returns a fully qualified name thatincludes the package, hash, and name, e.g.
+--
+-- "containers-0.5.6.2@conta_2C3ZI8RgPO2LBMidXKTvIU:Data.Map.Base.fromList".
+qualifiedName
+    :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => String -> Int -> Int -> String -> [String] -> m [String]
+qualifiedName targetModuleName lineNr colNr symbol importList = do
+        dflags <- GHC.getSessionDynFlags
+
+        setContext (map (IIDecl . simpleImportDecl . mkModuleName) (targetModuleName:importList))
+           `gcatch` (\(s  :: SourceError)    -> do gmLog GmDebug "qualifiedName"
```

Comment:
I don't get why you thought duplicating basically the same function three times here is a good idea. I just removed all of them except the one for SomeException. `show` on SomeException actually prints the value of the message for the real exception so there really is no need to say which exception type it was. Even if that was desired you can always check what type is inside `SomeException` using `Typeable`, so no need for this duplication.